src/primitives/fieldvec.rs

```diff
@@ -0,0 +1,380 @@
+//!
+//! This is very similar to the `ArrayVec` type from the `arrayvec` crate,
+//! with two major differences:
+//!
+//! * In the case that an allocator is available, switches to being unbounded.
+//! * It is specialized to field elements, and provides a number of utility
+//!   functions and constructors specifically for that case.
+//!
+//! Because it stores field elements, and fields always have a zero element,
+//! we can avoid working with uninitialized memory by setting "undefined"
+//! values to zero. There is theoretically a performance cost here, but
+//! given that our arrays are limited in size to low tens of elements, it
+//! is unlikely for this to be measurable.
+//!
+//! The purpose of this vector is to be a backing for the various (reduced)
+//! polynomials we encounter when processing BCH codes. These polynomials
+//! have degree <= the degree of the generator polynomial, whose degree
+//! in turn is a small integer (6 for bech32, 8 for descriptors, and 13
+//! or 15 for codex32, as examples).
+//!
+//! An example of a reduced polynomial is the residue computed when
+//! validating checksums. Typically, validating a BCH checksum just means
+//! computing this residue, comparing it to a target value, and throwing
+//! it away. However, we may want to keep the value in two cases:
+//!
+//! 1. When doing error correction, the residue value encodes the location
+//!    and values of the errors (assuming there are not too many).
+//! 2. When distinguishing between bech32 and bech32m, which differ only
+//!    in their target residues, we may want to know the computed residue
+//!    so we can do a manual comparison against both values.
+//!
+//! Despite these arrays being very small for all checksums we are aware
+//! of being practically used, in principle they can be any size, and we
+//! don't want to limit our users artificially. We cannot have arbitrary
+//! sized objects without an allocator, so we split the difference by
+//! using a fixed-size array, and when the user tries to go beyond this,
+//! panicking if an allocator is unavailable.
+//!
+//! Users of this type should take care not to expose this panic to users.
+//! This shouldn't be too hard, because this type is internal to the library
+//! which has two use cases:
+//!
+//! 1. Distinguishing bech32 and bech32m residues (within the limit).
+//! 2. Doing error correction (should have a small top-level API and easy
+//!    to early-detect things outside the limit and return an error).
+//!
+
+#[cfg(all(feature = "alloc", not(feature = "std")))]
+use alloc::vec::Vec;
+use core::{fmt, iter, mem, ops, slice};
+
+use super::Field;
+use crate::primitives::correction::NO_ALLOC_MAX_LENGTH;
+
+/// A vector of field elements.
+///
+/// Parameterized by the field type `F` which can be anything, but for most methods
+/// to be enabled needs `Default` and `Clone`. (Both are implied by `Field`.)
+#[derive(PartialEq, Eq, Clone, Debug, Hash)]
+pub struct FieldVec<F> {
```

Comment:
Why go to trouble of mentioning that this should not be exposed in the public API and make this type `pub`? Code seems to build with `pub(crate)`. Is it because of upcoming work?

Reply:
`pub(crate)` would make this visible to the entire crate. `pub` makes it visible only to the parent module (who in turn re-exports it privately to make it visible to other submodules).

I could do `pub(super)` if it makes you happier.

Reply:
Oh with private `field` module, I've never thought to do that before. Thanks for the explanation.